File: source/go/internal/jyotish/aspects.go
package jyotish

import (
	"sort"

	"github.com/ishankgupta95/panchang/source/go/v5/types"
)

var specialOffsets = [types.GrahaCount][]int{
	types.GrahaSun:     {},
	types.GrahaMoon:    {},
	types.GrahaMars:    {3, 7},
	types.GrahaMercury: {},
	types.GrahaJupiter: {4, 8},
	types.GrahaVenus:   {},
	types.GrahaSaturn:  {2, 9},
	types.GrahaRahu:    {},
	types.GrahaKetu:    {},
}

var node59Offsets = [2]int{4, 8}

func ComputeAspects(chart *types.BirthChart, options AspectsOptions) (types.AspectMap, error) {
	nodeAspects, err := resolveAspectsOptions(options)
	if err != nil {
		return types.AspectMap{}, err
	}

	var out types.AspectMap
	var seen [types.GrahaCount]int
	for i := range chart.Planets {
		p := &chart.Planets[i]
		if !p.Planet.Valid() {
			return types.AspectMap{}, types.NewPanchangError(
				"chart carries an out-of-range graha: "+itoa(int(p.Planet)),
				types.ErrInvalidInput)
		}
		seen[p.Planet]++
		out.SetForGraha(p.Planet, housesAspected(p, nodeAspects))
	}
	for _, g := range allGrahas {
		if seen[g] != 1 {
			return types.AspectMap{}, types.NewPanchangError(
				"chart must carry each graha exactly once; "+g.String()+" appears "+
					itoa(seen[g])+" times", types.ErrInvalidInput)
		}
	}
	return out, nil
}

func housesAspected(planet *types.PlanetPlacement, nodeAspects NodeAspects) []int {
	offsets := make([]int, 0, 5)
	offsets = append(offsets, 6) // universal 7th

	offsets = append(offsets, specialOffsets[planet.Planet]...)

	if nodeAspects == NodeAspects5And9 && planet.Planet.IsNode() {
		offsets = append(offsets, node59Offsets[:]...)
	}

	houses := make([]int, 0, len(offsets))
	var present [13]bool
	for _, o := range offsets {
		h := (planet.House-1+o)%12 + 1
		if h >= 1 && h <= 12 && !present[h] {
			present[h] = true
			houses = append(houses, h)
		}
	}
	sort.Ints(houses)
	return houses
}

File: source/ts/src/jyotish/aspects.ts
import type { AspectMap, BirthChart, GrahaName, PlanetPlacement } from '../types/jyotish';
import { PanchangError } from '../types/errors';

/** 0-based house offsets, beyond the universal 7th. BPHS Ch. 26. */
const SPECIAL_OFFSETS: Record<GrahaName, readonly number[]> = {
  Sun:     [],
  Moon:    [],
  Mars:    [3, 7],
  Mercury: [],
  Jupiter: [4, 8],
  Venus:   [],
  Saturn:  [2, 9],
  Rahu:    [],
  Ketu:    [],
};

/** Node 5th/9th aspects: BV Raman / KP convention, not BPHS. */
const NODE_5_9_OFFSETS: readonly number[] = [4, 8];

export interface AspectsOptions {
  /** `'7-only'` (default, BPHS-literal; so is `''`) or `'5-and-9'` (BV Raman / KP); anything else throws `INVALID_INPUT`. */
  nodeAspects?: '7-only' | '5-and-9';
}

/** Houses aspected by each graha: the universal 7th plus the special BPHS Ch. 26 aspects. */
export function computeAspects(
  chart: BirthChart,
  options: AspectsOptions = {},
): AspectMap {
  const nodeAspects = options.nodeAspects || '7-only';
  if (nodeAspects !== '7-only' && nodeAspects !== '5-and-9') {
    throw new PanchangError(
      `nodeAspects must be "7-only" or "5-and-9", got "${String(nodeAspects)}"`,
      'INVALID_INPUT',
    );
  }
  const out: Partial<Record<GrahaName, number[]>> = {};

  for (const planet of chart.planets) {
    out[planet.planet] = housesAspected(planet, nodeAspects);
  }

  return out as AspectMap;
}

function housesAspected(
  planet: PlanetPlacement,
  nodeAspects: '7-only' | '5-and-9',
): number[] {
  const offsets: number[] = [6];

  const special = SPECIAL_OFFSETS[planet.planet];
  for (const o of special) offsets.push(o);

  if (nodeAspects === '5-and-9' && (planet.planet === 'Rahu' || planet.planet === 'Ketu')) {
    for (const o of NODE_5_9_OFFSETS) offsets.push(o);
  }

  const houses = offsets.map((o) => ((planet.house - 1 + o) % 12) + 1);
  return Array.from(new Set(houses)).sort((a, b) => a - b);
}
